game_shared/ff/ff_gamerules.cpp

```diff
@@ -2087,27 +2087,45 @@ bool CFFGameRules::FCanTakeDamage( CBaseEntity *pVictim, CBaseEntity *pAttacker
 	// if it's a buildable, then we use the buildable's owner to perform the team checks etc. -> Defrag
 	CBasePlayer *pBuildableOwner = NULL;
 
+	bool isFriendlyFireOn = friendlyfire.GetInt() != 0;
 #ifdef GAME_DLL
 	// Special cases for sabotageable buildings
+	// Dexter 20181006: overhauled this section of code,
+	// it was a bit too sweeping, letting any sabotaged attackers or victims
+	// damage, or be damaged by anyone, regardless of friendly fire
+	// this fixes #327 , which was SG specific, but applied same
+	// treatment to detting dispensers
 
 	// If an SG is shooting its teammates then allow it to hurt them
 	if (pAttacker && pAttacker->Classify() == CLASS_SENTRYGUN)
 	{
 		CFFSentryGun *pSentry = dynamic_cast< CFFSentryGun* > (pAttacker);
 
-		if (pSentry && pSentry->IsMaliciouslySabotaged())
-			return true;
+		if ( pSentry && pSentry->IsMaliciouslySabotaged() )
+		{
+			bool victimIsSabTeammate = FFGameRules()->IsTeam1AlliedToTeam2( 
+				pVictim->GetTeamNumber(), 
+				pSentry->m_iSaboteurTeamNumber ) == GR_TEAMMATE;
+
+			return victimIsSabTeammate ? isFriendlyFireOn : true;
+		}
 	}
 	
 	if ( pVictim && pVictim->Classify() == CLASS_SENTRYGUN )
 	{
 		CFFSentryGun *pSentry = dynamic_cast <CFFSentryGun *> (pVictim);
 
 		// Allow team to kill their own SG if it is sabotaged
-		if (pSentry && pSentry->IsSabotaged())
-			return true;
+		if ( pSentry && pSentry->IsSabotaged() )

```

Comment:
Haven't thought this through fully, but it seems like it should be possible to condense most/all of these if blocks into a single block and treat all buildables the same as appropriate, right? Check if `Classify() == CLASS_SENTRYGUN || Classify() == CLASS_DISPENSER`, then just cast to `CFFBuildableObject` or w/e, and all the code inside the if could be the same.

Reply:
yes, strongly agree. I would have made it shorter but i didnt spend the time. Ill try to